src/main/java/org/apache/sysml/runtime/controlprogram/paramserv/LocalAggregationService.java

```diff
@@ -0,0 +1,151 @@
+		_funcName = keys[0];
+		if (keys.length == 2) {
+			_funcNS = keys[0];
+			_funcName = keys[1];
+		}
+		FunctionProgramBlock func = _ec.getProgram().getFunctionProgramBlock(_funcNS, _funcName);
+		_inputs = func.getInputParams();
+		_outputs = func.getOutputParams();
+		_boundInputs = _inputs.stream()
+				.map(input -> new CPOperand(input.getName(), input.getValueType(), input.getDataType()))
+				.toArray(CPOperand[]::new);
+		_inputNames = _inputs.stream().map(DataIdentifier::getName).collect(Collectors.toCollection(ArrayList::new));
+		_outputNames = _outputs.stream().map(DataIdentifier::getName).collect(Collectors.toCollection(ArrayList::new));
+	}
+
+	@Override
+	public void run() {
+
+		// Firstly, start by broadcasting the model
+		pushGlobalParams();
+		ParamservUtils.populate(_ec, _hyperParams, _inputs);
+		ParamservUtils.populate(_ec, _model, _inputs);
+
+		while (isAlive()) {
+			int finished = 0;
+			boolean[] tables = new boolean[_workerNum];
+
+			for (int i = 0; i < _workerNum && finished < _workerNum; i++) {
+				if (tables[i]) {
+					continue;
+				}
+				// Pull the gradients of each worker and update the model
+				Data gradients = _ps.pull(ParamServer.GRADIENTS_PREFIX + i);
+				if (gradients != null) {
+					LOG.info(String.format("Successfully pulled the gradients [size:%d kb] of worker_%d.",
+							((ListObject) gradients).getDataSize() / 1024, i));
+					aggregate((ListObject) gradients);
+					finished++;
+					tables[i] = true;
+				}
+				try {
+					TimeUnit.SECONDS.sleep(POLL_FREQUENCY);
+				} catch (InterruptedException e) {
+					throw new DMLRuntimeException("Aggregation service: Failed to sleep when polling the gradients.",
+							e);
+				}
+			}
+
+			// Push the updated model to ps
+			pushGlobalParams();
+
+		}
+
+		_ps.push(ParamServer.RESULT_MODEL, ParamservUtils.copyList(_model));
+
+		// Clean up the old model
+		ParamservUtils.cleanupListObject(_ec, _model);
+		ParamservUtils.cleanupListObject(_ec, _hyperParams);
+		_ec.getVariables().removeAll();
+	}
+
+	private void aggregate(ListObject gradients) {
+		// Populate the variables table with the gradients
+		ParamservUtils.populate(_ec, gradients, _inputs);
+
+		// Invoke the aggregate function
+		FunctionCallCPInstruction inst = new FunctionCallCPInstruction(_funcNS, _funcName, _boundInputs, _inputNames,
+				_outputNames, "aggregate function");
+		inst.processInstruction(_ec);
+
+		// Get the output
+		List<Data> paramsList = _outputs.stream().map(id -> _ec.getVariable(id.getName())).collect(Collectors.toList());
+
+		// Update the model with the new output
+		_model = new ListObject(paramsList, _outputNames);
+
+		// Clean up the gradients
+		ParamservUtils.cleanupListObject(_ec, gradients);
+	}
+
+	/**
+	 * Push the global parameters to ps
+	 */
+	private void pushGlobalParams() {
+		IntStream.range(0, _workerNum).forEach(i -> {
+			// Copy the model
+			ListObject copiedLO = ParamservUtils.copyList(_model);
```

Comment:
No need to copy anything here due to copy-on-write semantics - i.e., whenever a worker tries to update a model, we create a copy of it first.